src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                Sse.StoreScalar(dst + idx[i], x);
+            }
+        }
+
+        internal static void ScaleU(float scale, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* pdst = dst)
+                {
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        dstVector = Sse.Multiply(scaleVector, dstVector);
+                        Sse.Store(pDstCurrent, dstVector);
+
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        dstVector = Sse.MultiplyScalar(scaleVector, dstVector);
+                        Sse.StoreScalar(pDstCurrent, dstVector);
+
+                        pDstCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static void AddScaleU(float scale, float[] src, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        srcVector = Sse.Multiply(srcVector, scaleVector);
+                        dstVector = Sse.Add(dstVector, srcVector);
+                        Sse.Store(pDstCurrent, dstVector);
+
+                        pDstCurrent += 4;
+                        pSrcCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        srcVector = Sse.MultiplyScalar(srcVector, scaleVector);
+                        dstVector = Sse.AddScalar(dstVector, srcVector);
+                        Sse.StoreScalar(pDstCurrent, dstVector);
+
+                        pDstCurrent++;
+                        pSrcCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static void AddScaleSU(float scale, float[] src, int[] idx, float[] dst, int count)
```

Comment:
Does `S` here stand for `Sparse`, or something else?

Reply:
Yes, and `U` stands for unaligned.  See https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L5-L15.

We don't necessarily need to follow the same naming pattern here.  It may make sense to keep them aligned.  But it may be more confusing if/when the native assembly is removed.